### ISSUE TITLE
Do not include Assets in namespace for new classes

### ISSUE THREAD

**Issue by justinlueders** [JetBrains/resharper-unity#6]
Resharper wants to both modify namespaces to include the top level Assets folder as well as add Assets to the namespace of generated classes.  I don't think most people want this.

**Comment by alexander-mironov** [comment]
+1

**Comment by citizenmatt** [comment]
Could you give an example please? What's the folder structure, and what files are affected?

**Comment by justinlueders** [comment]
When creating a class via resharper, it will place it in a namespace according to the folder structure.  The root of the unity folder structure is always Assets.  So all namespaces start with that.  namespace Assets.Foo.Bar instead of the preferred Foo.Bar

**Comment by citizenmatt** [comment]
Cool. That sounds like we can mark the `Assets` root folder as not being a "namespace provider", and all the other folders will be used as part of the namespace instead. You can try this out by selecting the Assets folder, opening the Properties tool window and changing "Namespace Provider" to `false`. Does that solve the issue?

**Comment by justinlueders** [comment]
rad, sounds like it will, my 30 day trial of 2015 just ran out..need to
find my key for version 8 and ill test it. :)

**Comment by Shorebound** [comment]
I've checked the assets folder to false (all three projects link to the same folder, so when you change one all of them do), but this still doesn't work and I can't find a way around it.

I currently have a default namespace set as TeamName.GameName, but it's suggesting TeamName.GameName.Assets through ReSharper, even though there are no other instances of the Assets folder in the solution explorer. 
![output1](https://cloud.githubusercontent.com/assets/10943141/10839123/f22a1e86-7f23-11e5-9a53-7cf88cebd239.PNG)

**Comment by citizenmatt** [comment]
That's the approach you need to take, but I'm surprised it's not working. I've just tried it with a Unity VS2015 project, and ReSharper is correctly suggesting a namespace that contains the default namespace + folders, but not including "Assets". Do you have a screenshot or something that can help debug it? (A repro project would be great!)

**Comment by citizenmatt** [comment]
I'm taking a deeper look at this - what's the default behaviour for Unity? If I create a new project, I get standard assets that include scripts that have all sorts of different namespaces - `UnityStandardAssets._2D`, `UnityStandardAssets.Cameras`, `UnityStandardAssets.CrossPlatformInput.Inspector`, etc.

These don't match the file location - respectively `Assets\Standard Assets\2D\Scripts`, `Assets\Standard Assets\Cameras`, and `Assets\Editor\CrossPlatformInput`. This last one doesn't match at all.

Does Unity have a standard naming convention that we can apply? Or do people just build their own, and simply want `Assets` to be left out of it?

**Comment by Shorebound** [comment]
I would personally prefer to just build my own. I don't think Unity really have a standard naming convention.

I have been able to replicate the issue on my side and from what I can tell, I think the issue is with the multiple .csproj files that Unity generates. I have tried seeing if it's some strange inconsistency with multiple subfolders inside of the Assets folder being set to true, but that does not seem to be the case.

I just created a new Unity Project and uploaded it to [it's own repo](https://github.com/Shorebound/ResharperNamingIssue). Unity lets you set the root namespace in Edit -> Project Settings -> Editor. This sets the default namespace in VS to that namespace. THis only happens if the solution is not already created or if it's deleted as Unity regenerates the solution and project files every time there's a change.

In the repo is a Test.cs file in the Scripts folder. Both Assets and Scripts are set to false (or they are on my end - maybe it's some weird visual bug)

Test.cs' namespace defaults to TeamName.GameName.Assets. If that's the same for you when you download the project, the only way I've found to fix it so far is by actually deleting the Editor and Standard Assets folders (through Unity) in the Unity project which removes the other .csprojs. Once that is done, it rebuilds the solution with just the single .csproj, resetting the Assets folder back to True. Setting it to False now makes ReSharper behave correctly.

**Comment by citizenmatt** [comment]
Thanks for the repo, that helps. It looks like it's not set up right, though. Disabling the namespace provider sets a value in the `.csproj.dotSettings` files, and there's an entry for `Assets` in [`Plugins.csproj.dotSettings`](https://github.com/Shorebound/ResharperNamingIssue/blob/8ad303f3178e00f89431548d488e69423ae7c1fd/ReSharperNamingIssue.CSharp.Plugins.csproj.DotSettings#L2), so `Plugins` gets the correct namespace suggestion.

`Editor.csproj` doesn't have a file, so `Assets` is still treated as part of the suggested namespace, and `ResharperNamingIssue.CSharp.csproj.dotSettings` disables the namespace provider for `Assets\Scripts`, but *not* `Assets`. So, `Assets` is still part of the suggested namespace, but `Scripts` isn't.

You need to disable the namespace provider on `Assets` in all projects. Would you normally disable it for `Scripts` as well?

**Comment by Shorebound** [comment]
Yep, my own process is disabling `Assets` and `Scripts`. Then using subfolders under `Scripts` as part of the namespace. 

The main one that I want changed is the `CSharp.csproj` as that's where most of your code goes. 

The `Editor.csproj` is essentially for code that extends the Unity Editor itself and `Plugins.csproj` ... actually I have no clue what that one is for. The one in my main project, *Stranded*, is just filled with readmes from third party assets. I think Unity automatically searches for files underneath the `Editor` and `Plugins` folders in the project structure (no matter where they are) and then makes sure they're in their respective csprojects. It also recompiles the solution and .csprojects on every load.

In both the `ResharperNamingIssue` repo and my *Stranded* project, the `Assets` folder is disabled for each project. However, it doesn't let me change each project independently. If I set one `Assets` folder to `False`, they're all set to `False`. 

In *Stranded* (I can link you to a repo of it if necessary, but that project is a couple hundred mb, so I figured it would be easier to use a new one that replicates the problem) I don't even have a `.dotSettings` file for any `.csproj` but Visual Studio still states that the Assets folder is set to false. 

Is it possible to manually add in the `.dotSettings` file for each project or regenerate them somehow?

**Comment by citizenmatt** [comment]
I can't recreate the behaviour you're seeing - changing one `Assets` folder only affects that project, no others. And if you change the value, you should get a `.csproj.dotSettings` file. However, it looks like there is [a bug](https://youtrack.jetbrains.com/issue/RSRP-450738) - if you change the value, and then delete the file, the value doesn't reset back to the default of `true`. The data is cached in the solution cache as well as the `.csproj.dotSettings` file, and the cache doesn't seem to be updated when the file is deleted. I'd guess this is what's happened to you - if you reset the cache and reopened the solution, I'd expect the namespace provider settings to have reverted to `true`.

**Comment by citizenmatt** [comment]
I've been thinking about how to implement this, and one option is to automatically generate the `.csproj.dotSettings` file, but that file should ideally be checked in to source control, and I don't like the idea of changing the status of your project just by opening the solution.

So I've knocked up a quick prototype that will create an in-memory settings layer, effectively giving me a `.csproj.dotSettings` file in-memory, that I can programmatically write settings into (and it has a lower priority that `.csproj.dotSettings`, too, so you can easily override it). It looks like it's working nicely, and it will automatically create this settings layer for any Unity project you open. Currently, I'm disabling the `Assets` folder as a namespace provider.

Should I also disable `Assets\Scripts`?

Oh, and other settings can go in there too (I'm thinking specifically of C# language level - #5)

**Comment by Shorebound** [comment]
Hey, sorry for the late response! Exams have just started up.

Which cache is this that I would need to reset? The primary VS one? Might be a good idea to do it any way. 

Also, that seems like a great idea! If the option to disable Assets/Scripts is there it'd be much appreciated. I'll definitely give this a try

Thanks a lot for your help with this

**Comment by citizenmatt** [comment]
It's ReSharper's solution cache, which you can delete via ReSharper &rarr; Options &rarr; General. You'll need to close/reopen the solution for it to take effect.

**Comment by citizenmatt** [comment]
I've just pushed this update to the extension manager. It will stop `Assets` and `Scripts` being considered as part of the suggested namespace. Please update and give it a go!
